lib/_http_client.js

```diff
@@ -44,26 +44,26 @@ const errors = require('internal/errors');
 // with an additional rule for ignoring percentage-escaped characters, but
 // that's a) hard to capture in a regular expression that performs well, and
 // b) possibly too restrictive for real-world usage. So instead we restrict the
-// filter to just control characters and spaces.
+// filter to just control characters, spaces and two-byte characters.
 //
 // This function is used in the case of small paths, where manual character code
 // checks can greatly outperform the equivalent regexp (tested in V8 5.4).
 function isInvalidPath(s) {
   var i = 0;
-  if (s.charCodeAt(0) <= 32) return true;
+  if (s.charCodeAt(0) <= 32 || s.charCodeAt(0) > 0xFF) return true;
```

Comment:
Shouldn't this be `> 127` instead of `> 255` ? UTF-8 starts using multiple bytes after the last ASCII character (127).

Reply:
V8 strings are not UTF-8, but UTF-16 (UCS2). If it was UTF-8, this would not be necessary, as each byte in an UTF-8 multibyte sequence is `> 127` so it would be impossible to get ascii characters that way, even if some bytes were stripped.

What happens is that the string `' \uff2e'` is represented in node as `(uint16_t*){0xff2e}` (where `{}` is used here to denote an array). With the latin1 encoding (that's what gets used for headers), node internally calls `WriteOneByte` on that string to write it to the socket and `WriteOneByte` will strip the higher byte of the two-byte character, so we end up with `0x2e`. It works fine for all characters that can be represented with a single byte (and in UCS2, that means all characters below `0x100`) (whether any non-ascii characters make sense in this context is questionable, but it also doesn't break anything so I opted not to filter that).

Reply:
> With the latin1 encoding (that's what gets used for headers)

Unfortunately, it's not that simple.  See https://github.com/nodejs/node/issues/13296#issuecomment-305162413 for an explanation.  The two word summary is "it depends" (on the encoding of the body, but that doesn't fit in two words.)

Aside: calling `.charCodeAt()` twice for the same index is somewhat wasteful, it would be better to cache the result in a variable.  (V8's TurboFan compiler can inline the calls but I don't believe it will fold them.)

Reply:
Oh, thanks for pointing me to that issue. I thought a bit about something like that while writing the patch, but I couldn't find out where the encoding is actually set.

Reply:
At least as long as we aren't using chunked encoding (no idea what that is :))

Reply:
All code paths in `_http_outgoing.js` (except the chunked encoding one) seem to call `_send(something, 'latin1')` or `_send(something, null)` before any other calls to `_send`. I don't really understand the code flow, but that's what searching for `_send` tells me. So I may have missed something. Are non-latin1 headers only possible with chunked encoding?

Reply:
Scratch that, sorry for the spam. I just noticed that I swapped the branches of the `if` condition, the non-latin1 case happens when *not* using chunked encoding (which makes sense).

Reply:
An alternative: just set a flag in the constructor (like "latin1_safe") and then check it when sending the headers and throw if `latin1` gets used and `!latin1_safe` ? Would that work, or does that have other drawbacks, aside from being a little hacky?

Reply:
Sorry for the delay, missed your replies.  I don't know if your exact proposal would work but take a look at the `_storeHeader()` and `_send()` methods in `lib/_http_outgoing.js`, that's where the headers are put on the wire and their encoding is determined.

You may also want to look at the `_storeHeader()` calls in `lib/_http_client.js`, they compose the status line.

Reply:
This raises the question if we actually *want* to support non-latin1 encoded request path? Maybe `utf-8` makes sense in some contexts? I am not sure.

Reply:
I spent some time browsing through open issues yesterday and there's already a bug logged against this behaviour. (Would need to go back to find it again.) IMO what mscdex said above makes sense, just cap it at 127 instead. The current behaviour is completely unpredictable and broken.

Reply:
Here's the issue I was mentioning https://github.com/nodejs/node/issues/13296 — there's some good discussion there too.

I would prefer what I said above, at least for 9.x (maybe different solution for LTS), but I'll defer to @bnoordhuis and others with more experience with the `http` module.

Reply:
> This raises the question if we actually want to support non-latin1 encoded request path?  Maybe utf-8 makes sense in some contexts?

It's in fairly widespread use.  Rejecting it outright would almost certainly result in quite a bit of fallout and lots of frustrated users.  Rejecting malformed UTF-8 should be acceptable and appropriate, though.

(But as discussed, what to accept and what to reject depends on the encoding.)